rust/datafusion/src/test/mod.rs

```diff
@@ -135,6 +135,13 @@ pub fn format_batch(batch: &RecordBatch) -> Vec<String> {
             }
             let array = batch.column(column_index);
             match array.data_type() {
+                DataType::Utf8 => s.push_str(
+                    array
+                        .as_any()
+                        .downcast_ref::<array::StringArray>()
+                        .unwrap()
+                        .value(row_index),
+                ),
```

Comment:
See 15e6f0b.

Reply:
I am really sorry, I misread this. What I meant was not to test the `format_batch` function, but to test `count distinct` in integration.

However, I see that this was not the module I though it was. The code in this module is perfect as is now.

I meant adding a test [here](https://github.com/apache/arrow/blob/master/rust/datafusion/tests/sql.rs), which is where we perform integration tests.

Something along the lines of:

```rust
#[tokio::test]
async fn query_count_distinct() -> Result<()> {
    let schema = Arc::new(Schema::new(vec![
        Field::new("c1", DataType::Int32, true),
    ]));

    let data = RecordBatch::try_new(
        schema.clone(),
        vec![
            Arc::new(Int32Array::from(vec![Some(0), Some(1), None, Some(3), Some(3)])),
        ],
    )?;

    let table = MemTable::new(schema, vec![vec![data]])?;

    let mut ctx = ExecutionContext::new();
    ctx.register_table("test", Box::new(table));
    let sql = "SELECT COUNT(DISTINCT c1) FROM test";
    let actual = execute(&mut ctx, sql).await;
    let expected = vec![
        vec!["0", "1"],
        vec!["1", "1"],
        vec!["3", "2"],
    ];
    assert_eq!(expected, actual);
    Ok(())
}
```

does this make sense?

Reply:
No problem! 

This makes perfect sense, and I added the test (2d0999a). However, I also had to make the `count(distinct)` field nullable (4cdb951), otherwise I ran into an assertion error [here](https://github.com/apache/arrow/blob/2d0999ac7f1d01a4fe392add53720ceeb6b0b1f3/rust/datafusion/tests/sql.rs#L652) due to a mismatch of the field's nullability between the logical plan and physical plan. 

The logical plan nullability is set [here](https://github.com/apache/arrow/blob/2d0999ac7f1d01a4fe392add53720ceeb6b0b1f3/rust/datafusion/src/logical_plan/mod.rs#L298), and is always nullable. All of the regular aggregate expressions mark their field as nullable, for example, [here](https://github.com/apache/arrow/blob/2d0999ac7f1d01a4fe392add53720ceeb6b0b1f3/rust/datafusion/src/physical_plan/expressions.rs#L846) for the regular `Count`. 

I might be mistaken, but I think regular and distinct counts should be non-nullable? In any case, I went with making `count(distinct)` nullable for consistency with `count()`. Perhaps there's a follow-up here?

Reply:
I think that `count(distinct)` should be null when all entries are null, for consistency with `count`. I do not think that this blocks this, though. It is an edge case IMO.